self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoices.java

```diff
@@ -0,0 +1,277 @@
+        }
+        //create the invoice task.
+        final int taskInvoiceId = activeInvoiceId;
+        final InvoiceTask invoiceTask = new InvoiceTask() {
+            @Override
+            public int invoiceId() {
+                return taskInvoiceId;
+            }
+            @Override
+            public Duration timeSpent() {
+                return timeSpent;
+            }
+            @Override
+            public Task task() {
+                return task;
+            }
+        };
+        //add the task to invoice.
+        invoiceTasks
+            .computeIfAbsent(activeInvoiceId, k -> new ArrayList<>())
+            .add(invoiceTask);
+
+        return invoiceTask;
+    }
+
+    @Override
+    public boolean isPaid(final int id) {
+        return payedIds
+            .values()
+            .stream()
+            .anyMatch(v -> v.contains(id));
+    }
+
+    @Override
+    public List<InvoiceTask> tasks(final int id) {
+        return invoiceTasks.getOrDefault(id, List.of());
+    }
+
+    @Override
+    public Iterator<Invoice> iterator() {
+        //invoices across all contracts
+        return invoices
+            .values()
+            .stream()
+            .flatMap(Collection::stream)
+            .iterator();
+    }
+
+    /**
+     * Simulates "payment" of an invoice.
+     * Used here to aid better representation of
+     * {@link InMemoryInvoices} state (as in "deactivating" the invoice).
+     *
+     * @param id Invoice id.
+     */
+    public void pay(final int id) {
+        final Map.Entry<Contract.Id, Integer> foundActiveEntry = activeIds
+            .entrySet()
+            .stream()
+            .filter(e -> e.getValue() == id)
+            .findFirst()
+            .orElse(null);
+        if (foundActiveEntry == null) {
+            throw new IllegalStateException("There is no active invoice "
+                + " with id " + id);
+        }
+        final Contract.Id contractId = foundActiveEntry.getKey();
+        final Integer invoiceId = foundActiveEntry.getValue();
+        payedIds.computeIfAbsent(contractId, k -> new ArrayList<>())
+            .add(invoiceId);
+        activeIds.remove(contractId);
+    }
+
+    /**
+     * Contract Invoices delegate. Manages invoices for a contract.
+     */
+    private static final class ContractInvoices implements Invoices {
+        /**
+         * Id contract of this invoices.
+         */
+        private final Contract.Id contractId;
+
+        /**
+         * Storage context.
+         */
+        private final Storage storage;
+
+        /**
+         * Stored invoices for this contract.
+         * We use a stream because we don't want to load in
+         * memory all the contract invoices.
+         * <br>
+         * In order to "reuse" them, since streams are one time use only,
+         * we wrap the stream in a supplier, .
+         */
+        private final Supplier<Stream<Invoice>> contractInvoices;
```

Comment:
@criske Why do we need supplier of Stream? Can't we work with a List (actual filtering being done in an upper layer), like other examples such as ProjectTasks, ProjectContributors etc?

Reply:
If a contract has 1 million invoices then all of them will be loaded in mem, if we're using a list. Basically we will throw away the paging benefits from self-storage.
The stream will not emit any items until you "subscribe" to it.

Reply:
Maybe it is a little bit forced in this case.

Reply:
@criske But there will not be 1M invoices in the Contract, for sure. Same for Tasks. A Project cannot have more than, say 1000 tasks.

Think of each Task (and Invoice element) as of a Github issue. How many Github Projects do you know, which have 1M Issues? :D